packages/test/test-version-utils/src/describeWithVersions.ts

```diff
@@ -0,0 +1,148 @@
+        try {
+            await promise;
+        } catch (e) {
+            console.error(e);
+            hadErrors = true;
+        }
+    }
+
+    if (hadErrors) {
+        throw new Error("Exceptions while installing package versions. Check STDERR");
+    }
+};
+
+const defaultTimeoutMs = 20000;
+const defaultRequestedVersions: IRequestedFluidVersions = { requestRelativeVersions: -2 };
+
+function createTestSuiteWithInstalledVersion(
+    tests: (this: Mocha.Suite, provider: () => ITestObjectProvider) => void,
+    requiredVersions: IRequestedFluidVersions = defaultRequestedVersions,
+    timeoutMs: number = defaultTimeoutMs,
+) {
+    return function(this: Mocha.Suite) {
+        let defaultProvider: TestObjectProvider;
+        let resetAfterEach: boolean;
+        before(async function() {
+            this.timeout(Math.max(defaultTimeoutMs, timeoutMs));
+
+            await installRequiredVersions(requiredVersions);
+            defaultProvider = await getVersionedTestObjectProvider(
+                pkgVersion, // baseVersion
+                pkgVersion, // loaderVersion
+                {
+                    type: driver,
+                    version: pkgVersion,
+                    config: {
+                        r11s: { r11sEndpointName },
+                        odsp: { tenantIndex },
+                    },
+                }, // driverConfig
+                pkgVersion, // runtimeVersion
+                pkgVersion, // dataRuntimeVersion
+            );
+
+            Object.defineProperty(this, "__fluidTestProvider", { get: () => defaultProvider });
+        });
+
+        tests.bind(this)((options?: ITestObjectProviderOptions) => {
+            resetAfterEach = options?.resetAfterEach ?? true;
+            if (options?.syncSummarizer === true) {
+                defaultProvider.resetLoaderContainerTracker(true /* syncSummarizerClients */);
+            }
+
+            return defaultProvider;
+        });
+
+        afterEach(function(done: Mocha.Done) {
+            const logErrors = getUnexpectedLogErrorException(defaultProvider.logger);
+            // if the test failed for another reason
+            // then we don't need to check errors
+            // and fail the after each as well
+            if (this.currentTest?.state === "passed") {
+                done(logErrors);
+            } else {
+                done();
+            }
+
+            if (resetAfterEach) {
+                defaultProvider.reset();
+            }
+        });
+    };
+}
+
+type DescribeSuiteWithVersions =
+    (name: string,
+        tests: (
+            this: Mocha.Suite,
+            provider: (options?: ITestObjectProviderOptions) => ITestObjectProvider) => void
+    ) => Mocha.Suite | void;
+
+type DescribeWithVersions = DescribeSuiteWithVersions & Record<"skip" | "only", DescribeSuiteWithVersions>;
+
+/**
+ * Creates a test suite which will priorly install a set of requested Fluid versions for the tests to use.
```

Comment:
I think this should mention that using this function will ensure that the requested versions are installed before any of the test code runs. That is the reason for having a separate utitlity for this and not doing it directly in the tests right?

Reply:
That's right. Thank you for pointing it out. Will do.